### ISSUE TITLE
Bughunting: Run_WasmFloat32SubMinusZero and Run_WasmFloat64SubMinusZero

### ISSUE THREAD

**Issue by palfia** [v8mips/v8mips#314]
The Run_WasmFloat32SubMinusZero and Run_WasmFloat64SubMinusZero tests are failing on the targets:
https://build.chromium.org/p/client.v8.ports/builders/V8%20Mips%20-%20big%20endian%20-%20nosnap%20-%201/builds/481

https://v8.mips.com/bbv8/builders/Target%20-%20edge/builds/7185

Please investigate these failures.

**Comment by marijantic** [comment]
Addressed the comments. Implementation in https://codereview.chromium.org/2019693002/ returns a quiet NaN with the preserved payload. Now, tests on r2 simulator are failing, because quiet/signalling bits are different on the host platform and in mips. Implementing the fix to the simulator.

**Comment by kilvadyb** [comment]
The NaN handling on simulators is not trivial (on r2 sim when a C function is called then the C code needs std::numeric_limits<>::...nan which is the Intel NaN, not the MIPSr2 NaN but the code must work on r2 MIPS boards as well). That is why I suggested something like Heap::kNanValueRootIndex -- roots[kNanValue] -- which contains the correct (quiet?)nan value of the target architecture and solves r2, r6, sim problems. Although computing masks from it has some cost...

**Comment by marijantic** [comment]
The problem is that the heap value of NaN contains a canonical NaN with no payload. The failing tests expect that payload is preserved. Also, heap NaN is a double value, not a float, and all constants in HeapNumber class are meant to handle double NaN values.

**Comment by kilvadyb** [comment]
Then we should introduce a float heap NaN if float NaN is required. Looks like that for the payload we need a nan mask which should be computed in runtime for the architecture the v8 is running on.

**Comment by marijantic** [comment]
I'm afraid that introducing heap constants will not solve the problem. Because we need to check if an operand is any of the possible NaN bit patterns, and then we have to return the quiet version of the input NaN, and not the predefined value. 
I am not sure that the desired behavior is well defined, either. 
Wasm spec says the following: 

> > When the result of any arithmetic operation other than neg, abs, or copysign is a NaN, the sign bit and the fraction field (which does not include the implicit leading digit of the significand) of the NaN are computed as follows:
> 
> ```
> If the operation has any NaN input values, implementations may select any of them to be the result value, but with the most significant bit of the fraction field overwritten to be 1.
> 
> If the implementation does not choose to use an input NaN as a result value, or if there are no input NaNs, the result value has a nondeterministic sign bit, a fraction field with 1 in the most significant bit and 0 in the remaining bits.
> ```

On the other hand, code in wasm-interpreter suggests that some of the operations expect that the quiet NaN is returned, not the input operand "with the most significant bit of the fraction field overwritten to be 1". My first patch set was returning the NaN value as required by the specification text, but after examining the interpreter code, I switched to returning the quiet NaN. 
The requirement to preserve the payload while quieting the NaN, causes another r2 specific issue: simply changing the signaling bit is not enough, so payload cannot be preserved in all cases. For example, wasm-interpreter contains a function 

```
static float quiet(float a) {
  static const uint32_t kSignalingBit = 1 << 22;
  uint32_t q = bit_cast<uint32_t>(std::numeric_limits<float>::quiet_NaN());
  if ((q & kSignalingBit) != 0) {
    // On some machines, the signaling bit set indicates it's a quiet NaN.
    return bit_cast<float>(bit_cast<uint32_t>(a) | kSignalingBit);
  } else {
    // On others, the signaling bit set indicates it's a signaling NaN.
    return bit_cast<float>(bit_cast<uint32_t>(a) & ~kSignalingBit);
  }
}
```

The problem with this function is that it will not work well for all of the possible mips r2 NaN bit patterns (signaling NaN with only the signaling bit set will be converted to infinity). It does work well for the canonical NaNs, though, as these contain the payload of all ones.

**Comment by marijantic** [comment]
I switched to using heap NaN to determine the encoding in macro-assembler methods SubNanPreservePayloadAndSign_d, SubNanPreservePayloadAndSign_s (instead of checking FCSR). This solves the simulator issue, but it also means that we do not simulate r2 NaN type correctly (i.e. both r2 and r6 simulators trigger the same code path).

**Comment by marijantic** [comment]
Landed as #333
